syntax = "proto3";

package whatup;

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

option go_package = "github.com/digital-witness-lab/whatup/protos/whatupcore";


service WhatUpCore {
    rpc GetConnectionStatus(google.protobuf.Empty) returns (ConnectionStatus) {}
    rpc GetMessage(google.protobuf.Empty) returns (stream Message) {}
    rpc DownloadMedia(MediaLocator) returns (Media) {}
}

message JID {
    string user = 1;
    uint32 agent = 2;
    uint32 device = 3;
    uint32 server = 4;
    uint32 ad = 5;
}

message ConnectionStatus {
    bool connected = 1;
    google.protobuf.Timestamp timestamp = 2;
}

message Message {
    MessageInfo info = 1;
    MessageContent content = 2;
    MessageProperties messageProperties = 3;
}

message MessageSource {
    JID chat = 1;
    JID sender = 2;
    JID broadcastListOwner = 3;

    bool isFromMe = 4;
    bool isGroup = 5;
}

message MessageInfo {
    MessageSource source = 1;
    google.protobuf.Timestamp timestamp = 2;
    string id = 3;
    string pushName = 4;
    string type = 5; // TODO: turn into enum!

    string category = 6; // TODO: turn into enum!
    bool multicast = 7;
    //string mediaType = 8; moved into message content
}

message MessageContent {
    string text = 1;
    string mimeType = 2;
    MediaLocator mediaLocator = 3;
}

message MessageProperties {
    bool isEphemeral = 1;
    bool isViewOnce = 2;
    bool isDocumentWithCaption = 3;
    bool isEdit = 4;
}

message MediaLocator {
    map<string, bytes> locator = 1;
}

message Media {
    bytes Contents = 1;
}
